core/modules/qdisp/Executive.cc

```diff
@@ -275,80 +236,56 @@ void Executive::markCompleted(int jobId, bool success) {
-Executive::newNotifier(Executive &e, int jobId) {
-    return NotifyExecutive::newInstance(e, jobId);
-}
-
 void Executive::requestSquash(int jobId) {
-    RequesterPtr toSquash;
-    bool needToWarn = false;
-    ResponseRequester::Error e;
+    JobQuery::Ptr toSquash;
     {
-        std::lock_guard<std::mutex> lock(_requestersMutex);
-        RequesterMap::iterator i = _requesters.find(jobId);
-        if(i != _requesters.end()) {
-            e = i->second->getError();
-            if(e.isNone()) {
-                toSquash = i->second; // Remember which one to squash
-            } else {
-                needToWarn = true;
-            }
-        } else {
-            throw Bug("Executive::requestSquash() with invalid jobId");
+        std::lock_guard<std::mutex> lock(_jobsMutex);
+        auto iter = _jobMap.find(jobId);
+        if(iter == _jobMap.end()) {
+            LOGF_WARN("requesSquash invalid jobID %1%" % jobId);
+            return;
         }
+        toSquash = iter->second;
     }
-    if(needToWarn) { // Log outside the mutex
-        LOGF_WARN("Warning, requestSquash(jobId=%1%), but this job has already failed (%3%, %4%)."
-                  % jobId % e.getCode() % e.getMsg());
-    }
-
-    if(toSquash) { // Squash outside of the mutex
-        toSquash->cancel();
-    }
+    // release mutex before calling cancel.
+    toSquash->cancel();
 }
 
 void Executive::squash() {
-    bool alreadyCancelled = atomicSet(_cancelled, true, _cancelledMutex);
+    bool alreadyCancelled = _cancelled.set(true);
     if(alreadyCancelled) {
         LOGF_WARN("Executive::squash() already cancelled! refusing.");
         return;
     }
 
     LOGF_INFO("Trying to cancel all queries...");
-    std::vector<RequesterPtr> pendingRequesters;
-    std::ostringstream os;
-    os << "STATE=";
+    std::vector<JobQuery::Ptr> jobsToSquash;
     {
-        std::lock_guard<std::mutex> lock(_requestersMutex);
-        _printState(os);
-        RequesterMap::iterator i,e;
-        for(i=_requesters.begin(), e=_requesters.end(); i != e; ++i) {
-            pendingRequesters.push_back(i->second);
+        std::lock_guard<std::mutex> lock(_jobsMutex);
+        for(auto i=_jobMap.begin(), e=_jobMap.end(); i != e; ++i) {
+            jobsToSquash.push_back(i->second);
         }
     }
-    LOGF_INFO("%1%\nLoop cancel all queries..." % os.str());
-    LOGF_INFO("Enqueued requesters for cancelling...done");
+    LOGF_INFO("Enqueued jobs for cancelling...done");
     {
-        std::vector<RequesterPtr>::iterator i, e;
-        for(i=pendingRequesters.begin(), e=pendingRequesters.end(); i != e; ++i) {
+        for(auto i=jobsToSquash.begin(), e=jobsToSquash.end(); i != e; ++i) {
             // Could get stuck because it waits on xrootd,
             // which may be waiting on a thread blocked in _unTrack().
-            // Don't do this while holding _requestersMutex
-            (**i).cancel();
+            // Don't do this while holding _statusesMutex
```

Comment:
Good point. These are now operation on 2 different maps and 2 different mutex, so this extra bit shouldn't be needed.